quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -123,12 +127,24 @@ class
   nextState :: Typeable a => state -> Action state a -> Var a -> state
   nextState s _ _ = s
 
+  -- | Transition function for negative actions. Note that most negative testing applications
+  -- should not require an implementation of this function!
+  negativeNextState :: Typeable a => state -> Action state a -> state
+  negativeNextState s _ = s
+
   -- | Precondition for filtering generated `Action`.
   -- This function is applied before the action is performed, it is useful to refine generators that
   -- can produce more values than are useful.
   precondition :: state -> Action state a -> Bool
   precondition _ _ = True
 
+  -- | Precondition for filtering an `Action` that can meaningfully run but is supposed to fail.
+  -- An action will run as a negative action if it's precondition fails. This action should have
```

Comment:
`if it's precondition fails` --> `if it's negativePrecondition fails` seems more explicit, otherwise it quickly gets confusing.

Reply:
Also, perhaps makes it explicit that `precondition` has the precedence?